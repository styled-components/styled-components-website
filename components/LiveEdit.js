import React, { useEffect, useState } from 'react';
import styled, {
  createGlobalStyle,
  css,
  keyframes,
  withTheme,
  StyleSheetManager,
  ThemeProvider,
} from 'styled-components';
import stylisRTLPlugin from 'stylis-plugin-rtl';
import rem from '../utils/rem';
import { darkGrey, red } from '../utils/colors';
import { phone } from '../utils/media';
import { headerFont, monospace } from '../utils/fonts';

import '../utils/prismTemplateString';

import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live';

const StyledProvider = styled(LiveProvider)`
  box-shadow: ${rem(1)} ${rem(1)} ${rem(20)} rgba(20, 20, 20, 0.27);
  overflow: hidden;
  margin: ${rem(35)} 0;
  text-align: left;
`;

const Row = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: stretch;
  align-items: stretch;

  ${phone(css`
    flex-direction: column;
  `)};
`;

const columnMixin = css`
  flex-basis: 50%;
  width: 50%;
  max-width: 50%;

  ${phone(css`
    flex-basis: auto;
    width: 100%;
    max-width: 100%;
    height: auto;
  `)};
`;

export const editorMixin = `
  background: ${darkGrey};
  font-size: 0.8rem;
  font-family: ${monospace};
  font-weight: 300;
  height: ${rem(400)};
  overflow-y: auto !important;
  overflow-x: hidden;
  cursor: text;
  white-space: pre-wrap;
  position: relative;
`;

const StyledEditor = styled(LiveEditor)`
  ${editorMixin};
  ${columnMixin};
`;

const StyledPreview = styled(LivePreview)`
  position: relative;
  padding: 0.5rem;
  background: white;
  color: black;
  height: auto;
  overflow: hidden;
  white-space: normal;

  ${columnMixin};
`;

export const StyledError = styled(LiveError)`
  display: block;
  width: 100%;
  padding: ${rem(8)};
  background: ${red};
  color: white;
  font-size: 0.8rem;
  font-family: ${headerFont};
  white-space: pre;
`;

const LiveEdit = ({ noInline, code, scope = {} }) => {
  const [mounted, setMounted] = useState(false);

  useEffect(() => setMounted(true), []);

  return (
    <StyledProvider
      code={code}
      noInline={noInline}
      mountStylesheet={false}
      scope={{
        ...scope,
        createGlobalStyle,
        css,
        keyframes,
        styled,
        ThemeProvider,
        StyleSheetManager,
        withTheme,
        stylisRTLPlugin,
      }}
    >
      <Row>
        <StyledEditor />

        {/* because react-live uses a different babel compiler, the classnames it generates aren't stable and a remount is needed after SSR */}
        <StyledPreview key={mounted ? 'preview-client' : 'preview-ssr'} />
      </Row>

      <StyledError />
    </StyledProvider>
  );
};

export default LiveEdit;
