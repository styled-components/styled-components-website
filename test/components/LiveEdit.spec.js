import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';
import LiveEdit, { StyledError } from '../../components/LiveEdit';

test('LiveEdit renders correctly', () => {
  const wrapper = mount(<LiveEdit />);

  expect(wrapper).toMatchSnapshot();
});

test('StyledError renders correctly', () => {
  const tree = mount(<StyledError />, {
    context: { live: {} },
    childContextTypes: { live: PropTypes.object },
  });

  expect(tree).toMatchSnapshot();
});
