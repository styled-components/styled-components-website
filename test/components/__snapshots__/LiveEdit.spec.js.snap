// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveEdit renders correctly 1`] = `
.c0 {
  box-shadow: 0.05555555555555555rem 0.05555555555555555rem 1.1111111111111112rem rgba(20,20,20,0.27);
  overflow: hidden;
  margin: 1.9444444444444444rem 0;
  text-align: left;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c2 {
  background: #1d1f27;
  font-size: 0.8rem;
  font-family: "Operator Mono SSm A","Operator Mono SSm B",monospace;
  font-weight: 300;
  height: 22.22222222222222rem;
  overflow-y: scroll;
  overflow-x: hidden;
  cursor: text;
  white-space: pre-wrap;
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  width: 50%;
  max-width: 50%;
}

.c3 {
  position: relative;
  padding: 0.5rem;
  background: white;
  color: black;
  height: auto;
  overflow: hidden;
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  width: 50%;
  max-width: 50%;
}

.c4 {
  display: block;
  width: 100%;
  padding: 0.4444444444444444rem;
  background: #ff5555;
  color: white;
  font-size: 0.8rem;
  font-family: "Avenir Next",-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  white-space: pre;
}

@media (max-width:40.625em) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (max-width:40.625em) {
  .c2 {
    -webkit-flex-basis: auto;
    -ms-flex-preferred-size: auto;
    flex-basis: auto;
    width: 100%;
    max-width: 100%;
    height: auto;
  }
}

@media (max-width:40.625em) {
  .c3 {
    -webkit-flex-basis: auto;
    -ms-flex-preferred-size: auto;
    flex-basis: auto;
    width: 100%;
    max-width: 100%;
    height: auto;
  }
}

<LiveEdit>
  <Styled(LiveProvider)
    mountStylesheet={false}
    scope={
      Object {
        "ThemeProvider": [Function],
        "css": [Function],
        "keyframes": [Function],
        "styled": [Function],
        "withTheme": [Function],
      }
    }
  >
    <LiveProvider
      className="c0"
      code=""
      mountStylesheet={false}
      noInline={false}
      scope={
        Object {
          "ThemeProvider": [Function],
          "css": [Function],
          "keyframes": [Function],
          "styled": [Function],
          "withTheme": [Function],
        }
      }
    >
      <div
        className="react-live c0"
        scope={
          Object {
            "ThemeProvider": [Function],
            "css": [Function],
            "keyframes": [Function],
            "styled": [Function],
            "withTheme": [Function],
          }
        }
      >
        <styled.div>
          <div
            className="c1"
          >
            <Styled(LiveEditor)>
              <Styled(LiveEditor)>
                <LiveEditor
                  className="c2"
                >
                  <Editor
                    className="c2"
                    code=""
                    contentEditable={true}
                    language="jsx"
                    onChange={[Function]}
                  >
                    <pre
                      className="prism-code c2"
                      contentEditable={true}
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "",
                        }
                      }
                      onChange={[Function]}
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      spellCheck="false"
                    />
                  </Editor>
                </LiveEditor>
              </Styled(LiveEditor)>
            </Styled(LiveEditor)>
            <Styled(LivePreview)>
              <LivePreview
                className="c3"
              >
                <div
                  className="react-live-preview c3"
                />
              </LivePreview>
            </Styled(LivePreview)>
          </div>
        </styled.div>
        <Styled(LiveError)>
          <LiveError
            className="c4"
          >
            <div
              className="react-live-error c4"
            >
              SyntaxError: Unexpected token )
            </div>
          </LiveError>
        </Styled(LiveError)>
      </div>
    </LiveProvider>
  </Styled(LiveProvider)>
</LiveEdit>
`;

exports[`StyledError renders correctly 1`] = `
.c0 {
  display: block;
  width: 100%;
  padding: 0.4444444444444444rem;
  background: #ff5555;
  color: white;
  font-size: 0.8rem;
  font-family: "Avenir Next",-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  white-space: pre;
}

<Styled(LiveError)>
  <LiveError
    className="c0"
  />
</Styled(LiveError)>
`;
